File: projects/order-creation/core/src/main/java/com/robinfood/core/dtos/transactionrequestdto/FinalProductDTO.java
package com.robinfood.core.dtos.transactionrequestdto;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.robinfood.core.dtos.DeductionDTO;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.validation.Valid;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Positive;
import javax.validation.constraints.Size;
import java.io.Serializable;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class FinalProductDTO implements Serializable {

    private static final long serialVersionUID = 7650630725663371671L;

    @NotNull
    @Valid
    private FinalProductArticleDTO article;

    @NotNull
    @Valid
    private FinalProductBrandDTO brand;

    private FinalProductCategoryDTO category;

    @NotNull
    @Size
    @Valid
    private List<FinalProductDiscountDTO> discounts;

    @Builder.Default
    private List<DeductionDTO> deduction = new ArrayList<>();

    @NotNull
    @Size
    @Valid
    private List<GroupDTO> groups;

    @NotNull
    @Positive
    private Long id;

    @NotBlank
    private String image;

    @NotBlank
    private String name;

    @NotNull
    @Positive
    private BigDecimal price;

    @NotNull
    @Positive
    private Integer quantity;

    @NotNull
    @Size
    @Valid
    private List<RemovedPortionDTO> removedPortions;

    @NotNull
    @Valid
    private FinalProductSizeDTO size;

    private String sku;

    @Size
    @Valid
    private List<FinalProductTaxDTO> taxes;

    @NotNull
    @Positive
    private BigDecimal totalPrice;

    @Builder.Default
    private BigDecimal co2Total = BigDecimal.ZERO;

    /**
     * Multiplies the value of the final product by its quantity
     *
     * @return the total value of a final product
     */

    @JsonIgnore
    public BigDecimal getTotalDiscounts() {

        BigDecimal totalDiscount = discounts
                .stream()
                .filter(isProduct -> isProduct.getIsProductDiscount().equals(Boolean.TRUE))
                .map(FinalProductDiscountDTO::getValue)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        totalDiscount = totalDiscount.multiply(new BigDecimal(quantity));

        return totalPrice.subtract(totalDiscount);

    }
}

File: projects/posv2/core/src/main/kotlin/com/robinfood/core/dtos/transactionrequest/FinalProductDTO.kt
package com.robinfood.core.dtos.transactionrequest

import java.math.BigDecimal
import javax.validation.Valid
import javax.validation.constraints.Min
import javax.validation.constraints.NotBlank
import javax.validation.constraints.NotNull
import javax.validation.constraints.Positive
import javax.validation.constraints.Size

data class FinalProductDTO (

    @NotNull
    @Valid
    val article: ArticleDTO,

    @NotNull
    @Valid
    val brand: BrandDTO,

    @Min(0)
    val co2Total: BigDecimal?,

    @NotNull
    @Size
    @Valid
    val discounts: List<DiscountDTO>,

    @NotNull
    @Size
    @Valid
    val groups: List<GroupDTO>,

    @NotNull
    @Positive
    val id: Long,

    @NotBlank
    val image: String,

    @NotBlank
    val name: String,

    @NotNull
    @Positive
    val price: BigDecimal,

    @NotNull
    @Positive
    val quantity: Int,

    @NotNull
    @Size
    @Valid
    val removedPortions: List<RemovedPortionDTO>,

    @NotNull
    @Valid
    val size: SizeDTO,

    @NotBlank
   val sku: String,

    @Size
    @Valid
    val taxes: List<TaxDTO>,

    @NotNull
    @Positive
    val totalPrice: BigDecimal
)